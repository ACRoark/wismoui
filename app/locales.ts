export const DEFAULT_LOCALE = 'en';

const PROD_LOCALES = {
  en: 'English',
};

const TEST_LOCALES = {
  en: 'English',
  es: 'Spanish',
};

export const SUPPORTED_LOCALES = process.env.NODE_ENV === 'production' ? PROD_LOCALES : TEST_LOCALES;
