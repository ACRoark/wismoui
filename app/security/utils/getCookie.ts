const getCookie = (name: string): string =>
  document.cookie.split('; ').reduce((previousValue: string, currentValue: string): string => {
    const parts = currentValue.split('=');

    return parts[0] === name ? decodeURIComponent(parts[1]) : previousValue;
  }, '');

export default getCookie;
