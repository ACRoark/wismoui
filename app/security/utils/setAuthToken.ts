import setCookie from './setCookie';

const setAuthToken = (value?: string): void => {
  if (value) {
    setCookie('token', `Bearer ${value}`, { days: 1, path: '/' });
  }
};

export default setAuthToken;
