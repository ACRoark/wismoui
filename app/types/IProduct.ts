import ProductType from './ProductType';

interface IProduct {
  docId?: string | null;
  isCanceled?: boolean;
  processedAtDate?: string | null;
  productType: ProductType;
  serviceDate: string;
}

export default IProduct;
