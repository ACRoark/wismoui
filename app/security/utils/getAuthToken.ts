import getCookie from './getCookie';

const getAuthToken = (): string => getCookie('token');

export default getAuthToken;
