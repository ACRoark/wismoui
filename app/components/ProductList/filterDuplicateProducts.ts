import { IProduct } from 'types';

const filterDuplicateProducts = (products: IProduct[]): IProduct[] => {
  const cache = {};
  const filteredProducts: IProduct[] = [];

  for (let i = products.length - 1; i >= 0; i--) {
    if (!cache[products[i].productType]) {
      cache[products[i].productType] = true;
      filteredProducts.push(products[i]);
    }
  }

  return filteredProducts;
};

export default filterDuplicateProducts;
