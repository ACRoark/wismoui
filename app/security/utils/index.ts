import getAuthToken from './getAuthToken';
import setAuthToken from './setAuthToken';
import userIsAuthenticated from './userIsAuthenticated';

export { getAuthToken, setAuthToken, userIsAuthenticated };
