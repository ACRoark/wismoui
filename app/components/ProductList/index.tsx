import * as React from 'react';

import ProductName from 'components/ProductName';
import { IProduct, ProductType, ProductTypes } from 'types';
import filterDuplicateProducts from './filterDuplicateProducts';

interface IProductListProps {
  products: IProduct[] | null;
}

const isKnown = (productType: ProductType): boolean => ProductTypes.includes(productType);

const ProductList: React.FC<IProductListProps> = (props: IProductListProps): React.ReactElement => {
  const { products } = props;

  if (!products) {
    return <></>;
  }

  let productList;

  const filteredProducts = products.filter(
    (product: IProduct): boolean => !product.isCanceled && isKnown(product.productType),
  );
  const filteredDuplicates = filterDuplicateProducts(filteredProducts);
  productList = filteredDuplicates.map(
    (product: IProduct, index: number): React.ReactElement => (
      <React.Fragment key={product.productType}>
        <ProductName productType={product.productType}/>
        {index < filteredDuplicates.length - 1 ? ', ' : null}
      </React.Fragment>
    ),
  );

  return <>{productList}</>;
};

export default ProductList;
