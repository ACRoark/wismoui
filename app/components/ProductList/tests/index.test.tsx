import { SUPPORTED_LOCALES } from 'locales';
import React from 'react';
import { Provider } from 'react-redux';

import ProductList from 'components/ProductList';
import { createMockStore, createSnapshotWithIntl, renderWithIntl } from 'testing/utils';
import IProduct from 'types/IProduct';

const store = createMockStore();

const emptyList: IProduct[] =[];
const oneItemList: IProduct[] = [
  {
    productType: 'EFR_D1_1',
    serviceDate: '2020-03-21',
  },
];
const threeItemList: IProduct[] = [
  {
    productType: 'EFR_D1_1',
    serviceDate: '2020-03-21',
  },
  {
    productType: 'GCR_A_H',
    serviceDate: '2020-03-15',
  },
  {
    productType: 'EFR_D1',
    serviceDate: '04-22-2020',
  },
];

const expectNoConsoleErrors = (language: string, products: IProduct[]): void => {
  const spy = jest.spyOn(global.console, 'error');

  renderWithIntl(<Provider store={store}><ProductList products={products} /></Provider>, language);

  expect(spy).not.toHaveBeenCalled();
};

const testProductListSnapshot = (language: string, products: IProduct[] | null): void => {
  const tree = createSnapshotWithIntl(<Provider store={store}><ProductList products={products} /></Provider>, language);

  expect(tree).toMatchSnapshot();
};

describe('<ProductList />', (): void => {
  const languages = Object.keys(SUPPORTED_LOCALES);

  languages.forEach((language: string): void => {
    it(`should not log errors in console when rendered with a single item and the current language is ${language}`, (): void =>
      expectNoConsoleErrors(language, emptyList));

    it(`should not log errors in console when rendered with an empty list and the current language is ${language}`, (): void =>
      expectNoConsoleErrors(language, emptyList));

    it(`should not log errors in console when rendered with multiple items and the current language is ${language}`, (): void =>
      expectNoConsoleErrors(language, emptyList));

    it(`should render correctly when a single item is passed and the current language is ${language}`, (): void =>
      testProductListSnapshot(language, oneItemList));

    it(`should render correctly when an empty list is passed and the current language is ${language}`, (): void =>
      testProductListSnapshot(language, emptyList));

    it(`should render correctly when multiple items are passed and the current language is ${language}`, (): void =>
      testProductListSnapshot(language, threeItemList));

    it(`should render correctly when null is passed and the current language is ${language}`, (): void =>
      testProductListSnapshot(language, null));
  });
});
