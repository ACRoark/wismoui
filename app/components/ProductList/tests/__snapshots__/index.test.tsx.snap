// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductList /> should render correctly when a single item is passed and the current language is en 1`] = `"Residential Electric Interruptible Air Conditioning"`;

exports[`<ProductList /> should render correctly when a single item is passed and the current language is es 1`] = `"Aire acondicionado eléctrico interrumpible residencial"`;

exports[`<ProductList /> should render correctly when an empty list is passed and the current language is en 1`] = `null`;

exports[`<ProductList /> should render correctly when an empty list is passed and the current language is es 1`] = `null`;

exports[`<ProductList /> should render correctly when multiple items are passed and the current language is en 1`] = `
Array [
  "Residential Electric Service",
  ", ",
  "Natural Gas Residential Heating",
  ", ",
  "Residential Electric Interruptible Air Conditioning",
]
`;

exports[`<ProductList /> should render correctly when multiple items are passed and the current language is es 1`] = `
Array [
  "Servicio Eléctrico Residencial",
  ", ",
  "Calefacción Residencial de Gas Natural",
  ", ",
  "Aire acondicionado eléctrico interrumpible residencial",
]
`;

exports[`<ProductList /> should render correctly when null is passed and the current language is en 1`] = `null`;

exports[`<ProductList /> should render correctly when null is passed and the current language is es 1`] = `null`;
